features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/OpennmsKafkaProducer.java

```diff
@@ -408,6 +409,10 @@ private void processKafkaSendQueue() {
                     producer.send(producerRecord, (recordMetadata, e) -> {
                         if (e != null) {
                             LOG.warn("Failed to send record to producer: {}.", producerRecord, e);
+                            if(e instanceof TimeoutException) {
+                                // If Kafka is Offline, buffer the record again.
+                                kafkaSendQueue.offer(kafkaRecord);
```

Comment:
I'm assuming this ends up reordering the queue as it will be put at the back of the queue right, is that ok for our case?

Also since we are `offer()`ing the message here, we may still lose the message if the queue is full right now, is that ok as well?

Reply:
That's tradeoff I think worth debating. We probably need to make a choice.

Reply:
IMO not dropping events/alarms may be better solution. The trade-off is out of order events/alarms. Given that Kafka also doesn't guarantee order, think this is okay since most likely when Kafka gets connected, they will be delivered immediately small difference in delivery time may not matter. 

Also considered having synchronous delivery which would keep the order. But we have dependency on starting AlarmStore based on callback from events/alarms delivery. May not be worth tinkering that.

Reply:
@mattixtech  With new change, Now producer sends messages synchronously which would allow us to keep the order.

Reply:
Alarms should remain in order and Kafka does guarantee ordering for the same key/partition. The alarm topic was built so that we could consume the topic and build a KTable for the track the state of alarms in external systems - changing the order breaks this.

Re-ordering is okay for other payloads such as events & flows though.

Reply:
Got it.